src/main/java/io/confluent/idesidecar/restapi/resources/OAuthCallbackResource.java

```diff
@@ -109,15 +114,40 @@ public Uni<String> callback(@QueryParam("code") String authorizationCode,
     }
   }
 
+  /**
+   * Returns the URI that the callback page should redirect to for a specific CCloud connection.
+   * Attempts to read the URI from the connection's spec. If null, returns the URI set via the
+   * application config <code>ide-sidecar.connections.ccloud.oauth.vscode-extension-uri</code>.
+   *
+   * @param connection the CCloud connection
+   * @return the URI
+   */
+  private String getRedirectUri(CCloudConnectionState connection) {
+    var connectionSpec = connection.getSpec();
+
+    if (connectionSpec != null) {
+      var ccloudConfig = connectionSpec.ccloudConfig();
+      if (ccloudConfig != null && ccloudConfig.ideAuthCallbackUri() != null) {
+        return ccloudConfig.ideAuthCallbackUri();
+      }
+    }
+
+    return CCLOUD_OAUTH_VSCODE_EXTENSION_URI;
+  }
+
   private Future<String> renderFailure(Throwable error) {
+    return renderFailure(error, null);
+  }
+
+  private Future<String> renderFailure(Throwable error, String redirectUri) {
     Log.error(error);
     var errorMessage = error instanceof SSLHandshakeException
         ? TLS_HANDSHAKE_ERROR_MESSAGE
         : error.getMessage();
     return Future.succeededFuture(
         callbackFailure
             .data("error", errorMessage)
-            .data("vscode_redirect_uri", CCLOUD_OAUTH_VSCODE_EXTENSION_URI)
+            .data("vscode_redirect_uri", Objects.requireNonNullElse(redirectUri, ""))
```

Comment:
The callback page might be called with a non-CCloud connection or an invalid `state` param, in which case we don't have access to the connection-defined redirect URI.

@shouples In this case, do you want the sidecar to refrain from redirecting the user or do you want the sidecar to redirect the user to the redirect URI specified via the application config `ide-sidecar.connections.ccloud.oauth.vscode-extension-uri`?

Reply:
To close out this comment based on our offline discussion yesterday: we decided it made sense to not redirect the user in this case, since it could become confusing if the connection initially came from one product/distribution but something happened to cause the `state` param to become invalid and then the user would possibly be prompted with a different product/distribution redirect.